### PR TITLE
[BXMSPROD-2019] Productized branch creation

### PR DESCRIPTION
**Pull request by lampajr**
**JIRA:** https://issues.redhat.com/browse/BXMSPROD-2019

**Referenced Pull Requests:**
- https://github.com/kiegroup/kogito-pipelines/pull/967
- https://github.com/kiegroup/optaplanner/pull/2845
- https://github.com/kiegroup/drools/pull/5209

deploy artifacts configurable
- https://github.com/kiegroup/drools/pull/5285
- https://github.com/kiegroup/optaplanner/pull/2866

Improvements:
- Update build chain configuration on DSL config repository
- Update prod branch on all repos
- Downgrade Quarkus

### CODE REVIEW

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare [hunk 1]
```diff
@@ -146,27 +180,107 @@ pipeline{
...
+            steps{
+                script {
+                    def prodRepositories = getProdRepositoriesToBranch()
+                    prodRepositories.each { repoConfig ->
+                        String repository = repoConfig.name
+                        String repoAuthor = repoConfig.author?.name ?: readMainBranchConfig().git.author.name
+                        String repoCredsId = repoConfig.author?.credentials_id ?: readMainBranchConfig().git.author.credentials_id
+
+                        // TODO is this value saved somewhere such that we do not have to hardcode?
+                        String jobsGroovyPath = '.ci/jenkins/dsl/jobs.groovy'
```
Comment (lampajr) [inline dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]: where could we save this path in order to avoid hardcoding it? maybe we can simply save it on the branch.yaml, wdyt?

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare [hunk 2]
```diff
@@ -146,27 +180,107 @@ pipeline{
...
+                            // here productized branches should have been already created on repos to be productized
+                            checkout(githubscm.resolveRepository(repository, repoAuthor, getProductizedBranchFromRepository(repository), false, repoCredsId))
+
+                            getOrCreateGitBranch(repoConfig.branch, getRepositoryAuthorCredsId())
+
+                            if (fileExists("${jobsGroovyPath}")) {
+                                // def jobsGroovy = readFile file: jobsGroovyPath
+                                // remove fullProfile
+                                sh "sed -i 's/, addFullProfileJobParamsGetter)/)/g' ${jobsGroovyPath}"
+                                // TODO: disable pr checks in getMultijobPRConfig, get them from params
```
Comment (lampajr) [inline dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare:206]: This should be the last remaining part to implement, disable PR checks from `jobs.groovy` file.
I would do that in a separate PR as I need to re-think a bit what is the best way to handle it.

**Comment by lampajr** [comment]
`Jenkins DSL / Drools DSL Downstream tests - lampajr/drools (pull_request)` is failing for this:
```
org.kie.jenkins.jobdsl.JobScriptsSpec > test script /tmp/tmp.kqlIuPlzrE/dsl/seed/jobs/jobs.groovy FAILED
    org.spockframework.runtime.UnallowedExceptionThrownError at JobScriptsSpec.groovy:76
        Caused by: javaposse.jobdsl.dsl.DslScriptException at JobScriptsSpec.groovy:72
            Caused by: groovy.lang.MissingMethodException at JobScriptsSpec.groovy:72
```

**Comment by lampajr** [comment]
> `Jenkins DSL / Drools DSL Downstream tests - lampajr/drools (pull_request)` is failing for this:
> 
> ```
> org.kie.jenkins.jobdsl.JobScriptsSpec > test script /tmp/tmp.kqlIuPlzrE/dsl/seed/jobs/jobs.groovy FAILED
>     org.spockframework.runtime.UnallowedExceptionThrownError at JobScriptsSpec.groovy:76
>         Caused by: javaposse.jobdsl.dsl.DslScriptException at JobScriptsSpec.groovy:72
>             Caused by: groovy.lang.MissingMethodException at JobScriptsSpec.groovy:72
> ```

looks like it was a temporary/random issue :/ it was using forked drools branch that was not existing anymore (maybe cached?) - now all green

**Comment by lampajr** [comment]
Linked https://github.com/kiegroup/optaplanner/pull/2845 to setup DSL changes on OptaPlanner repo too

File: dsl/seed/jobs/root_jobs.groovy [hunk 3]
```diff
@@ -107,10 +107,19 @@ if (communityReleaseBranches) {
...
+            }
+
+            if (DEPENDENCY_PROJECTS) {
+                DEPENDENCY_PROJECTS.split(',').each { projectName ->
+                    stringParam("${projectName}_UPDATE_VERSION".toUpperCase(), '', "${Utils.getRepoNameCamelCase(projectName)} dependency version which this will depend on")
+                }
             }
 
             stringParam('QUARKUS_VERSION', '', 'Quarkus version to which to update all productized branches, usually latest LTS version')
+            stringParam('DOWNGRADE_PR_BRANCH', '', 'Which PR branch name to use for Quarkus downgrade? If none given, a name will be generated automatically.')
```
Comment (radtriste) [inline dsl/seed/jobs/root_jobs.groovy]: could be renamed to `DOWNGRADE_QUARKUS_PR_BRANCH` ?

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare [hunk 4]
```diff
@@ -133,6 +136,38 @@ pipeline{
                         sh "rm -f ${branchConfigFilePath}"
                         writeYaml file: "${branchConfigFilePath}", data: branchConfig, overwrite: true
 
+                        // update GHA build-chain configuration
+                        // comment out projects we are not interested in from the buildchain-config.yaml
+                        def projectToRemoveFromBCPRChecks = getProjectsToRemoveFromBuildChainConfig()
+
+                        if (projectToRemoveFromBCPRChecks && projectToRemoveFromBCPRChecks.size() > 0) {
```
Comment (radtriste) [inline dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]:
```suggestion
                        if (projectToRemoveFromBCPRChecks) {
```

should be enough in Groovy

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare [hunk 5]
```diff
@@ -133,6 +136,38 @@ pipeline{
                         sh "rm -f ${branchConfigFilePath}"
                         writeYaml file: "${branchConfigFilePath}", data: branchConfig, overwrite: true
 
+                        // update GHA build-chain configuration
```
Comment (radtriste) [inline dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]:
```suggestion
                        // update build-chain configuration
```

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare (hunk 1 above)
Comment (radtriste) [reply dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]: in the env ?

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare (hunk 2 above)
Comment (radtriste) [reply dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare:206]: I think we need to redo the way the PR checks are generated in those repositories, to make it more dynamic, based on:
- build-chain-config
- or a file in `.ci` folder (name to be considered)

we can discuss about that if you want

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare [hunk 6]
```diff
@@ -146,27 +181,136 @@ pipeline{
...
+                                    // if quarkusDowngradePrBranch has been created by update-quarkus-all job, checkout it
+                                    if (checkoutGitBranchIfExists(quarkusDowngradePrBranch, repoCredsId)) {
+                                        // apply patches from .ci/environments/quarkus-lts/patches
+                                        def patchesDirectory = '.ci/environments/quarkus-lts/patches'
+                                        def patches = findFiles(glob: "${patchesDirectory}/*")
+
+                                        if (patches && patches.size() > 0) {
+                                            patches.each { patch ->
+                                                echo "Applying patch ${patch}"
+                                                sh "git apply ${patch}"
```
Comment (radtriste) [inline dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]:
```suggestion
                                                sh "git apply --whitespace=fix ${patch}"
```

safety measure :)

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare [hunk 7]
```diff
@@ -146,27 +181,136 @@ pipeline{
...
+                                            }
+                                        } else {
+                                            echo "No patches found at ${patchesDirectory}"
+                                        }
```
Comment (radtriste) [inline dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]: You could just call `.ci/environments/update.sh quarkus-lts` to apply all changes. the script will take care of applying the patches :)

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare (hunk 6 above)
Comment (lampajr) [reply dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]: I agree :)

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare (hunk 7 above)
Comment (lampajr) [reply dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]: @radtriste yeah but calling `update.sh` should also trigger https://github.com/kiegroup/drools/blob/main/.ci/environments/common/update_quarkus.sh, right ? which is something that should be managed by the Quarkus upgrade job.
TBH they should be idempotent so even if I run again nothing should change. Do you think is it safer to run it?

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare (hunk 2 above)
Comment (lampajr) [reply dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare:206]: I agree as at the moment to disable them we should manipulate the file and it could be tricky to fully automate, furthermore changes on the file could also easily break the automation as it would be. So I agree on rethink how to setup those jobs.

File: dsl/seed/jobs/root_jobs.groovy (hunk 3 above)
Comment (lampajr) [reply dsl/seed/jobs/root_jobs.groovy]: done :)

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare (hunk 7 above)
Comment (radtriste) [reply dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]: I would say, better to do it twice than none ? it should not have any impact or ?

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare (hunk 1 above)
Comment (lampajr) [reply dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]: I decided to put it in the `branch.yaml > jenkins > jobs_definition_file` [1] to make it more customizable, wdyt?

[1] https://github.com/kiegroup/kogito-pipelines/pull/967/files#diff-e528bf449969213a2b24315837341a4cbe03d928d66ade752bdab2a628b6424eR144

File: dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare (hunk 7 above)
Comment (lampajr) [reply dsl/seed/jenkinsfiles/Jenkinsfile.prod.prepare]: based on the script it should not so I agree, doing it

**Review by radtriste — APPROVED** [review]

**Review by radtriste — APPROVED** [review]
